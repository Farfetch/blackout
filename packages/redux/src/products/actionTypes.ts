/**
 * @module products/actionTypes
 * @category Products
 * @subcategory Actions types
 */

// Details

/** Action type dispatched when the dehydrate product details occurs. */
export const DEHYDRATE_PRODUCT_DETAILS =
  '@farfetch/blackout-redux/DEHYDRATE_PRODUCT_DETAILS';

/** Action type dispatched when the fetch color grouping request fails. */
export const FETCH_PRODUCT_COLOR_GROUPING_FAILURE =
  '@farfetch/blackout-redux/FETCH_PRODUCT_COLOR_GROUPING_FAILURE';
/** Action type dispatched when the fetch color grouping request starts. */
export const FETCH_PRODUCT_COLOR_GROUPING_REQUEST =
  '@farfetch/blackout-redux/FETCH_PRODUCT_COLOR_GROUPING_REQUEST';
/** Action type dispatched when the fetch color grouping request succeeds. */
export const FETCH_PRODUCT_COLOR_GROUPING_SUCCESS =
  '@farfetch/blackout-redux/FETCH_PRODUCT_COLOR_GROUPING_SUCCESS';

/** Action type dispatched when the fetch measurements request fails. */
export const FETCH_PRODUCT_MEASUREMENTS_FAILURE =
  '@farfetch/blackout-redux/FETCH_PRODUCT_MEASUREMENTS_FAILURE';
/** Action type dispatched when the fetch measurements request starts. */
export const FETCH_PRODUCT_MEASUREMENTS_REQUEST =
  '@farfetch/blackout-redux/FETCH_PRODUCT_MEASUREMENTS_REQUEST';
/** Action type dispatched when the fetch measurements request succeeds. */
export const FETCH_PRODUCT_MEASUREMENTS_SUCCESS =
  '@farfetch/blackout-redux/FETCH_PRODUCT_MEASUREMENTS_SUCCESS';

/** Action type dispatched when the fetch product attributes request fails. */
export const FETCH_PRODUCT_ATTRIBUTES_FAILURE =
  '@farfetch/blackout-redux/FETCH_PRODUCT_ATTRIBUTES_FAILURE';
/** Action type dispatched when the fetch product attributes request starts. */
export const FETCH_PRODUCT_ATTRIBUTES_REQUEST =
  '@farfetch/blackout-redux/FETCH_PRODUCT_ATTRIBUTES_REQUEST';
/** Action type dispatched when the fetch product attributes request succeeds. */
export const FETCH_PRODUCT_ATTRIBUTES_SUCCESS =
  '@farfetch/blackout-redux/FETCH_PRODUCT_ATTRIBUTES_SUCCESS';

/** Action type dispatched when the fetch product details request fails. */
export const FETCH_PRODUCT_DETAILS_FAILURE =
  '@farfetch/blackout-redux/FETCH_PRODUCT_DETAILS_FAILURE';
/** Action type dispatched when the fetch product details request starts. */
export const FETCH_PRODUCT_DETAILS_REQUEST =
  '@farfetch/blackout-redux/FETCH_PRODUCT_DETAILS_REQUEST';
/** Action type dispatched when the fetch product details request succeeds. */
export const FETCH_PRODUCT_DETAILS_SUCCESS =
  '@farfetch/blackout-redux/FETCH_PRODUCT_DETAILS_SUCCESS';

/** Action type dispatched when the fetch product merchants locations request
 * fails.
 */
export const FETCH_PRODUCT_MERCHANTS_LOCATIONS_FAILURE =
  '@farfetch/blackout-redux/FETCH_PRODUCT_MERCHANTS_LOCATIONS_FAILURE';
/** Action type dispatched when the fetch product merchants locations request
 * starts.
 */
export const FETCH_PRODUCT_MERCHANTS_LOCATIONS_REQUEST =
  '@farfetch/blackout-redux/FETCH_PRODUCT_MERCHANTS_LOCATIONS_REQUEST';
/** Action type dispatched when the fetch product merchants locations request
 * succeeds.
 */
export const FETCH_PRODUCT_MERCHANTS_LOCATIONS_SUCCESS =
  '@farfetch/blackout-redux/FETCH_PRODUCT_MERCHANTS_LOCATIONS_SUCCESS';

/** Action type dispatched when the fetch product size guides request fails. */
export const FETCH_PRODUCT_SIZEGUIDES_FAILURE =
  '@farfetch/blackout-redux/FETCH_PRODUCT_SIZEGUIDES_FAILURE';
/** Action type dispatched when the fetch product size guides request starts. */
export const FETCH_PRODUCT_SIZEGUIDES_REQUEST =
  '@farfetch/blackout-redux/FETCH_PRODUCT_SIZEGUIDES_REQUEST';
/** Action type dispatched when the fetch product size guides request succeeds. */
export const FETCH_PRODUCT_SIZEGUIDES_SUCCESS =
  '@farfetch/blackout-redux/FETCH_PRODUCT_SIZEGUIDES_SUCCESS';

/** Action type dispatched when the fetch product sizes request fails. */
export const FETCH_PRODUCT_SIZES_FAILURE =
  '@farfetch/blackout-redux/FETCH_PRODUCT_SIZES_FAILURE';
/** Action type dispatched when the fetch product sizes request starts. */
export const FETCH_PRODUCT_SIZES_REQUEST =
  '@farfetch/blackout-redux/FETCH_PRODUCT_SIZES_REQUEST';
/** Action type dispatched when the fetch product sizes request succeeds. */
export const FETCH_PRODUCT_SIZES_SUCCESS =
  '@farfetch/blackout-redux/FETCH_PRODUCT_SIZES_SUCCESS';

/** Action type dispatched when the reset details entities occurs. */
export const RESET_PRODUCT_DETAILS_ENTITIES =
  '@farfetch/blackout-redux/RESET_PRODUCT_DETAILS_ENTITIES';
/** Action type dispatched when the reset details state occurs. */
export const RESET_PRODUCT_DETAILS_STATE =
  '@farfetch/blackout-redux/RESET_PRODUCT_DETAILS_STATE';

/** Action type dispatched when the fetch product fittings request fails. */
export const FETCH_PRODUCT_FITTINGS_FAILURE =
  '@farfetch/blackout-redux/FETCH_PRODUCT_FITTINGS_FAILURE';
/** Action type dispatched when the fetch product fittings request starts. */
export const FETCH_PRODUCT_FITTINGS_REQUEST =
  '@farfetch/blackout-redux/FETCH_PRODUCT_FITTINGS_REQUEST';
/** Action type dispatched when the fetch product fittings request succeeds. */
export const FETCH_PRODUCT_FITTINGS_SUCCESS =
  '@farfetch/blackout-redux/FETCH_PRODUCT_FITTINGS_SUCCESS';

// Products lists: listing and sets

/** Action type dispatched when the dehydrate products list occurs - used on
 * fetchListing and fetchSet actions. */
export const DEHYDRATE_PRODUCTS_LIST =
  '@farfetch/blackout-redux/DEHYDRATE_PRODUCTS_LIST';

/** Action type dispatched when the fetch products list request fails - used on
 * fetchListing and fetchSet actions. */
export const FETCH_PRODUCTS_LIST_FAILURE =
  '@farfetch/blackout-redux/FETCH_PRODUCTS_LIST_FAILURE';
/** Action type dispatched when the fetch products list request starts - used on
 * fetchListing and fetchSet actions. */
export const FETCH_PRODUCTS_LIST_REQUEST =
  '@farfetch/blackout-redux/FETCH_PRODUCTS_LIST_REQUEST';
/** Action type dispatched when the fetch products list request succeeds - used on
 * fetchListing and fetchSet actions. */
export const FETCH_PRODUCTS_LIST_SUCCESS =
  '@farfetch/blackout-redux/FETCH_PRODUCTS_LIST_SUCCESS';

/** Action type dispatched when the reset products list entities occurs - used on
 * `products.lists` reducer. */
export const RESET_PRODUCTS_LISTS_ENTITIES =
  '@farfetch/blackout-redux/RESET_PRODUCTS_LISTS_ENTITIES';
/** Action type dispatched when the reset products list state occurs - used on
 * fetchListing, fetchSet, reset products lists state actions and `products.lists` reducer. */
export const RESET_PRODUCTS_LISTS_STATE =
  '@farfetch/blackout-redux/RESET_PRODUCTS_LISTS_STATE';

/** Action type dispatched when the products list hash is setted - used on
 * fetchListing and fetchSet actions. */
export const SET_PRODUCTS_LIST_HASH =
  '@farfetch/blackout-redux/SET_PRODUCTS_LIST_HASH';

/** Action type dispatched when the fetch recommended set request fails. */
export const FETCH_RECOMMENDED_SET_FAILURE =
  '@farfetch/blackout-redux/FETCH_RECOMMENDED_SET_FAILURE';
/** Action type dispatched when the fetch recommended set request starts.  */
export const FETCH_RECOMMENDED_SET_REQUEST =
  '@farfetch/blackout-redux/FETCH_RECOMMENDED_SET_REQUEST';
/** Action type dispatched when the fetch recommended set request succeeds. */
export const FETCH_RECOMMENDED_SET_SUCCESS =
  '@farfetch/blackout-redux/FETCH_RECOMMENDED_SET_SUCCESS';
