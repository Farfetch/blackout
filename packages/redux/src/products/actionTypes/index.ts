export * from './productAttributes.actionTypes';
export * from './productColorGrouping.actionTypes';
export * from './productDetails.actionTypes';
export * from './productFittings.actionTypes';
export * from './productGrouping.actionTypes';
export * from './productMeasurements.actionTypes';
export * from './productMerchantsLocations.actionTypes';
export * from './productSizeGuides.actionTypes';
export * from './productSizes.actionTypes';
export * from './productsLists.actionTypes';
export * from './recentlyViewedProducts.actionTypes';
export * from './recommendedSet.actionTypes';
