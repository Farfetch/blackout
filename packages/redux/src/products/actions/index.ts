/**
 * Products actions.
 */

export * from './factories/index.js';

export { default as fetchProductListingFacets } from './fetchProductListingFacets.js';
export { default as fetchProductAttributes } from './fetchProductAttributes.js';
export { default as fetchProductDetails } from './fetchProductDetails.js';
export { default as fetchProductFittings } from './fetchProductFittings.js';
export { default as fetchProductGrouping } from './fetchProductGrouping.js';
export { default as fetchProductGroupingProperties } from './fetchProductGroupingProperties.js';
export { default as fetchProductListing } from './fetchProductListing.js';
export { default as fetchProductMeasurements } from './fetchProductMeasurements.js';
export { default as fetchProductOutfits } from './fetchProductOutfits.js';
export { default as fetchProductSet } from './fetchProductSet.js';
export { default as fetchProductSizeGuides } from './fetchProductSizeGuides.js';
export { default as fetchProductSizes } from './fetchProductSizes.js';
export { default as fetchProductVariantsByMerchantsLocations } from './fetchProductVariantsByMerchantsLocations.js';
export { default as fetchRecentlyViewedProducts } from './fetchRecentlyViewedProducts.js';
export { default as fetchRecommendedProducts } from './fetchRecommendedProducts.js';
export { default as fetchRecommendedProductSet } from './fetchRecommendedProductSet.js';
export { default as removeRecentlyViewedProduct } from './removeRecentlyViewedProduct.js';
export { default as resetProductDetails } from './resetProductDetails.js';
export { default as resetProductListings } from './resetProductListings.js';
export { default as saveRecentlyViewedProduct } from './saveRecentlyViewedProduct.js';
