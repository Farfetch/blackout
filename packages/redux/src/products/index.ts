export * as productsActionTypes from './actionTypes/index.js';

export { default as productsServerInitialState } from './serverInitialState/index.js';

export * from './actions/index.js';
export * from './selectors/index.js';
export * from './types/index.js';
export * from './utils/index.js';

export {
  default as productsReducer,
  entitiesMapper as productsEntitiesMapper,
} from './reducer/index.js';
