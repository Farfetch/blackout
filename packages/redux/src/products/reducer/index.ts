import { combineReducers } from 'redux';
import attributesReducer from './attributes';
import detailsReducer, {
  entitiesMapper as detailsEntitiesMapper,
} from './details';
import fittingsReducer from './fittings';
import groupingPropertiesReducer from './groupingProperties';
import groupingReducer from './grouping';
import listsReducer, { entitiesMapper as listsEntitiesMapper } from './lists';
import measurementsReducer, {
  entitiesMapper as measurementsEntitiesMapper,
} from './measurements';
import outfitsReducer from './outfits';
import recentlyViewedProductsReducer from './recentlyViewedProducts';
import recommendedProductsReducer from './recommendedProducts';
import recommendedSetsReducer from './recommendedSet';
import sizeGuidesReducer from './sizeGuides';
import sizesReducer from './sizes';
import variantsByMerchantsLocationsReducer from './variantsByMerchantsLocations';

export const entitiesMapper = {
  ...detailsEntitiesMapper,
  ...listsEntitiesMapper,
  ...measurementsEntitiesMapper,
};

const reducers = combineReducers({
  attributes: attributesReducer,
  grouping: groupingReducer,
  groupingProperties: groupingPropertiesReducer,
  details: detailsReducer,
  fittings: fittingsReducer,
  outfits: outfitsReducer,
  lists: listsReducer,
  measurements: measurementsReducer,
  recentlyViewed: recentlyViewedProductsReducer,
  recommendedProducts: recommendedProductsReducer,
  recommendedSets: recommendedSetsReducer,
  sizeGuides: sizeGuidesReducer,
  sizes: sizesReducer,
  variantsByMerchantsLocations: variantsByMerchantsLocationsReducer,
});

export default reducers;
