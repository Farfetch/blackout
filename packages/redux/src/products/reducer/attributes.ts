import * as actionTypes from '../actionTypes';
import { AnyAction, combineReducers } from 'redux';
import type {
  FetchProductAttributesAction,
  FetchProductAttributesFailureAction,
  FetchProductAttributesRequestAction,
  ProductsAttributesState,
} from '../types';

export const INITIAL_STATE: ProductsAttributesState = {
  error: {},
  isLoading: {},
};

const error = (
  state = INITIAL_STATE.error,
  action:
    | FetchProductAttributesRequestAction
    | FetchProductAttributesFailureAction,
) => {
  switch (action.type) {
    case actionTypes.FETCH_PRODUCT_ATTRIBUTES_REQUEST:
      return {
        ...state,
        [action.meta.productId]: undefined,
      };
    case actionTypes.FETCH_PRODUCT_ATTRIBUTES_FAILURE:
      return {
        ...state,
        [action.meta.productId]: action.payload.error,
      };
    default:
      return state;
  }
};

const isLoading = (
  state = INITIAL_STATE.isLoading,
  action: FetchProductAttributesAction | AnyAction,
) => {
  switch (action.type) {
    case actionTypes.FETCH_PRODUCT_ATTRIBUTES_REQUEST:
      return {
        ...state,
        [action.meta.productId]: true,
      };
    case actionTypes.FETCH_PRODUCT_ATTRIBUTES_SUCCESS:
      return { ...state, [action.meta.productId]: false };
    case actionTypes.FETCH_PRODUCT_ATTRIBUTES_FAILURE:
      return { ...state, [action.meta.productId]: undefined };
    default:
      return state;
  }
};

export const getError = (
  state: ProductsAttributesState,
): ProductsAttributesState['error'] => state.error;
export const getIsLoading = (
  state: ProductsAttributesState,
): ProductsAttributesState['isLoading'] => state.isLoading;

/**
 * Reducer for products attributes.
 *
 * @param state  - Current redux state.
 * @param action - Action dispatched.
 *
 * @returns New state.
 */
const productsAttributesReducer = combineReducers({
  error,
  isLoading,
});

export default productsAttributesReducer;
