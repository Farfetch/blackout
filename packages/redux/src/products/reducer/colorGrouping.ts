/**
 * @module products/reducer
 * @category Products color grouping
 * @subcategory Reducer
 */
import * as actionTypes from '../actionTypes';
import { AnyAction, combineReducers } from 'redux';
import type {
  FetchProductColorGroupingAction,
  FetchProductColorGroupingFailureAction,
  FetchProductColorGroupingRequestAction,
  ProductsColorGroupingState,
} from '../types';

export const INITIAL_STATE: ProductsColorGroupingState = {
  error: {},
  isLoading: {},
};

const error = (
  state = INITIAL_STATE.error,
  action:
    | FetchProductColorGroupingRequestAction
    | FetchProductColorGroupingFailureAction,
) => {
  switch (action.type) {
    case actionTypes.FETCH_PRODUCT_COLOR_GROUPING_REQUEST:
      return {
        ...state,
        [action.meta.productId]: undefined,
      };
    case actionTypes.FETCH_PRODUCT_COLOR_GROUPING_FAILURE:
      return {
        ...state,
        [action.meta.productId]: action.payload.error,
      };
    default:
      return state;
  }
};

const isLoading = (
  state = INITIAL_STATE.isLoading,
  action: FetchProductColorGroupingAction | AnyAction,
) => {
  switch (action.type) {
    case actionTypes.FETCH_PRODUCT_COLOR_GROUPING_REQUEST:
      return {
        ...state,
        [action.meta.productId]: true,
      };
    case actionTypes.FETCH_PRODUCT_COLOR_GROUPING_SUCCESS:
      return { ...state, [action.meta.productId]: false };
    case actionTypes.FETCH_PRODUCT_COLOR_GROUPING_FAILURE:
      return { ...state, [action.meta.productId]: undefined };
    default:
      return state;
  }
};

export const getError = (
  state: ProductsColorGroupingState,
): ProductsColorGroupingState['error'] => state.error;
export const getIsLoading = (
  state: ProductsColorGroupingState,
): ProductsColorGroupingState['isLoading'] => state.isLoading;

/**
 * Reducer for products color grouping.
 *
 * @static
 *
 * @param {object} state - Current redux state.
 * @param {object} action - Action dispatched.
 *
 * @returns {object} New state.
 */
const productsColorGroupingReducer = combineReducers({
  error,
  isLoading,
});

export default productsColorGroupingReducer;
