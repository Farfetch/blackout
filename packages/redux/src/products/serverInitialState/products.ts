import { INITIAL_STATE as DETAILS_INITIAL_STATE } from '../reducer/details';
import { normalize } from 'normalizr';
import { INITIAL_STATE as SIZES_INITIAL_STATE } from '../reducer/sizes';
import get from 'lodash/get';
import isEmpty from 'lodash/isEmpty';
import productSchema from '../../entities/schemas/product';
import type { Model, StoreState } from '../../types';

/**
 * Converts server data for details of a product to store state.
 *
 * @memberof module:products/reducer
 *
 * @param {object} data - Params injected by the server.
 * @param {object} data.model - Page model with product details data.
 * @param {object} data.options - General options for any modification.
 * @param {object} data.options.productImgQueryParam - Query parameter to be
 * appended to each product image URL.
 *
 * @returns {object} Initial state for the product details reducer.
 */
const serverInitialState = ({
  model,
  options: { productImgQueryParam } = {},
}: {
  model: Model;
  options?: { productImgQueryParam?: string };
}): {
  details: StoreState['products']['details'];
  entities?: StoreState['entities'];
  sizes: StoreState['products']['sizes'];
} => {
  // Check if a model object is of a product detail page (type === 'Product')
  // - if not, do nothing
  if (isEmpty(model) || get(model, 'dataLayer.general.type') !== 'Product') {
    return {
      details: DETAILS_INITIAL_STATE,
      sizes: SIZES_INITIAL_STATE,
    };
  }

  // Only passing specifically the properties that we see fit,
  // instead of spreading everything and end up with unnecessary data that
  // will only increase the payload/store.
  const {
    breadCrumbs,
    colorSet,
    colorSwatch,
    complementaryInformation,
    imageGroups,
    liveModel,
    price,
    productAttributes,
    productRef,
    productSize,
    recommendedSet,
    relatedSets,
    result: productData,
    scaleId,
    sizes,
    sizeSet,
    slug,
  } = model;
  const productInfo = {
    // Send this to the entity's `adaptProductImages`
    productImgQueryParam,
    // Data needed to support product information
    attributes: productAttributes,
    breadCrumbs,
    colorSet,
    colorSwatch,
    complementaryInformation,
    imageGroups,
    liveModel,
    price,
    productRef,
    productSize,
    recommendedSet,
    relatedSets,
    // @ts-expect-error scaleId is duplicated on product payload
    scaleId,
    sizes,
    sizeSet,
    slug,
    ...productData,
    // The model already brings a `measurements` entry,
    // which is different from the
    // '/api/products/{id}/variantsMeasurements' response.
    // Thus, we need to override it since it'll always be requested async.
    measurements: undefined,
  };
  const { result: id, entities } = normalize(productInfo, productSchema);

  return {
    details: {
      error: {},
      id,
      isHydrated: {
        [id]: true,
      },
      isLoading: {
        [id]: false,
      },
    },
    entities,
    sizes: {
      error: {},
      isLoading: {
        [id]: false,
      },
    },
  };
};

export default serverInitialState;
