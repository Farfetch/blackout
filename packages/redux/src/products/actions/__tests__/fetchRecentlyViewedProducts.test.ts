import { actionTypes } from '../../';
import { expectedRecentlyViewedRemotePayload } from 'tests/__fixtures__/products';
import { fetchRecentlyViewedProducts } from '../';
import { getRecentlyViewedProducts } from '@farfetch/blackout-client/products';
import { mockStore } from '../../../../tests';
import find from 'lodash/find';
import reducer from '../../reducer';
import type { StoreState } from '../../../types';

jest.mock('@farfetch/blackout-client/products', () => {
  return {
    ...jest.requireActual('@farfetch/blackout-client/products'),
    getRecentlyViewedProducts: jest.fn(),
  };
});

const mockAction = { type: 'this_is_a_mock_action' };

const mockRecentlyViewedStore = (state: StoreState = {}) =>
  mockStore(
    {
      products: {
        recentlyViewed: reducer(undefined, mockAction),
      },
    },
    state,
  );

describe('fetchRecentlyViewedProducts() action creator', () => {
  let store: ReturnType<typeof mockRecentlyViewedStore>;
  const expectedConfig = undefined;
  const query = {
    page: 1,
    pageSize: 25,
  };

  beforeEach(() => {
    jest.clearAllMocks();
    store = mockRecentlyViewedStore();
  });

  it('should create the correct actions for when the fetch recently viewed products fail', async () => {
    const expectedError = new Error('fetch recently viewed products error');

    (getRecentlyViewedProducts as jest.Mock).mockRejectedValueOnce(
      expectedError,
    );

    expect.assertions(4);

    await fetchRecentlyViewedProducts(query)(
      store.dispatch,
      store.getState as () => StoreState,
    ).catch(error => {
      expect(error).toBe(expectedError);
      expect(getRecentlyViewedProducts).toHaveBeenCalledTimes(1);
      expect(getRecentlyViewedProducts).toHaveBeenCalledWith(
        query,
        expectedConfig,
      );
    });

    expect(store.getActions()).toEqual(
      expect.arrayContaining([
        {
          type: actionTypes.FETCH_RECENTLY_VIEWED_PRODUCTS_REQUEST,
        },
        {
          type: actionTypes.FETCH_RECENTLY_VIEWED_PRODUCTS_FAILURE,
          payload: { error: expectedError },
        },
      ]),
    );
  });

  it('should create the correct actions for when the fetch recently viewed products procedure is successful', async () => {
    console.warn = jest.fn();

    (getRecentlyViewedProducts as jest.Mock).mockResolvedValueOnce(
      expectedRecentlyViewedRemotePayload,
    );

    const result = await fetchRecentlyViewedProducts(query, expectedConfig)(
      store.dispatch,
      store.getState as () => StoreState,
    );
    const actionResults = store.getActions();

    expect(result).toBe(expectedRecentlyViewedRemotePayload);
    expect(getRecentlyViewedProducts).toHaveBeenCalledTimes(1);
    expect(getRecentlyViewedProducts).toHaveBeenCalledWith(
      query,
      expectedConfig,
    );
    expect(actionResults).toEqual([
      {
        type: actionTypes.FETCH_RECENTLY_VIEWED_PRODUCTS_REQUEST,
      },
      {
        type: actionTypes.FETCH_RECENTLY_VIEWED_PRODUCTS_SUCCESS,
        payload: expectedRecentlyViewedRemotePayload,
      },
    ]);
    expect(
      find(actionResults, {
        type: actionTypes.FETCH_RECENTLY_VIEWED_PRODUCTS_SUCCESS,
      }),
    ).toMatchSnapshot('fetch recently viewed products success payload');

    store = mockRecentlyViewedStore({
      products: {
        recentlyViewed: {
          result: {
            remote: {
              entries: [{ productId: 123123, lastVisitDate: '10000' }],
              totalItems: 1,
              totalPages: 1,
              number: 1,
            },
            computed: undefined,
            pagination: undefined,
          },
          error: undefined,
          isLoading: false,
        },
      },
    } as StoreState);

    await fetchRecentlyViewedProducts(query, expectedConfig)(
      store.dispatch,
      store.getState as () => StoreState,
    );

    expect(console.warn).toHaveBeenCalledWith(`
              @farfetch/blackout-redux/products - Seems you are trying to fetch recently viewed products more than once.
              Please make sure you only request the products once, and use the "saveRecentlyViewedProduct" action to mark the product as viewed when a product page is visited.
              Keep in mind that "saveRecentlyViewedProduct" will only store locally the recently viewed product and will not persist it on the server.
              For that, make sure you are using analytics with Omnitracking integration.
          `);
  });
});
