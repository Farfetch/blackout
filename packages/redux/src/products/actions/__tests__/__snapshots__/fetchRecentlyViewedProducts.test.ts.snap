// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchRecentlyViewedProducts() action creator should create the correct actions for when the fetch recently viewed products procedure is successful: fetch recently viewed products success payload 1`] = `
Object {
  "payload": Object {
    "entries": Array [
      Object {
        "lastVisitDate": "2020-02-01T15:57:30.238Z",
        "productId": 22222222,
      },
      Object {
        "lastVisitDate": "2020-02-02T15:57:30.238Z",
        "productId": 11111111,
      },
    ],
    "number": 1,
    "totalItems": 1,
    "totalPages": 1,
  },
  "type": "@farfetch/blackout-redux/FETCH_RECENTLY_VIEWED_PRODUCTS_SUCCESS",
}
`;
